components/content/content-editor/content-editor.service.ts

```diff
@@ -0,0 +1,185 @@
+import { Fragment, Mark, Node, NodeType } from 'prosemirror-model';
+import { EditorState, Transaction } from 'prosemirror-state';
+import { EditorView } from 'prosemirror-view';
+import { imageMimeTypes, isImage } from '../../../utils/image';
+import { uuidv4 } from '../../../utils/uuid';
+
+export class ContentEditorService {
+	/**
+	 * Ensures that the last node in the editor doc is a specific node.
+	 */
+	public static ensureEndNode(tr: Transaction, nodeType: NodeType) {
+		if (tr.doc.lastChild && tr.doc.lastChild.type.name !== nodeType.name) {
+			const newNode = nodeType.create();
+			return tr.insert(tr.doc.nodeSize - 2, newNode);
+		}
+		return undefined;
+	}
+
+	public static getSelectedNode(state: EditorState) {
+		let selFrom = state.selection.from;
+		let node = state.doc.nodeAt(selFrom);
+		if (node === null && selFrom > 0) {
+			selFrom--;
+			node = state.doc.nodeAt(selFrom);
+		}
+		if (node === undefined) {
+			return null;
+		}
+		return node;
+	}
+
+	private static findParentNode(proposedParent: Node, child: Node): Node | null {
+		for (let i = 0; i < proposedParent.childCount; i++) {
+			const parentChild = proposedParent.child(i);
+			if (parentChild === child) {
+				return proposedParent;
+			} else {
+				const maybeParent = this.findParentNode(parentChild, child);
+				if (maybeParent !== null) {
+					return maybeParent;
+				}
+			}
+		}
+
+		return null;
+	}
+
+	public static getParentNode(state: EditorState, child: Node) {
+		return this.findParentNode(state.doc, child);
+	}
+
+	/**
+	 * Returns a list of all applied to any node within the selection
+	 */
+	public static getSelectionMarks(state: EditorState) {
+		const markTypes: Mark[] = [];
+		state.doc.nodesBetween(
+			state.selection.from,
+			state.selection.to,
+			(node: Node, _0: number, _1: Node, _2: number) => {
+				for (const mark of node.marks) {
+					if (!markTypes.some(m => m.type.name === mark.type.name)) {
+						markTypes.push(mark);
+					}
+				}
+			}
+		);
+		return markTypes;
+	}
+
+	/**
+	 * Indicates whether the given node is contained in a node of the given type.
+	 * Returns the parent node that matched, or `false`.
+	 */
+	public static isContainedInNode(state: EditorState, node: Node, nodeType: NodeType) {
+		if (node.type.name === nodeType.name) {
+			return node;
+		}
+
+		let child = node;
+		let parent = this.getParentNode(state, node);
+		while (parent !== null) {
+			if (parent.type.name === nodeType.name) {
+				return parent;
+			}
+
+			// Walk up
+			child = parent;
+			parent = this.getParentNode(state, child);
+		}
+		return false;
+	}
+
+	public static handleImageUploads(view: EditorView, items: DataTransferItemList) {
+		let handled = false;
+
+		for (let i = 0; i < items.length; i++) {
+			const transferItem = items[i];
+
+			if (
+				transferItem.kind === 'file' &&
+				imageMimeTypes.includes(transferItem.type.toLowerCase())
+			) {
```

Comment:
getAsFile might come at a performance cost, I think this is reasonable code duplication, since isImage only checks the imageMimeTypes array, which we import from there.